### DISCUSSION TITLE
Can netfox work for moba-style games?

### DISCUSSION THREAD

**Discussion by cody3b** [foxssake/netfox#214] [General]
The game I'm wanting to make will essentially have moba style combat, where precision and quick reactions are necessary. Is netfox suited for this? Attacks will primarily be projectile based, with aoe abilities as well.

**Comment by albertok** [comment]
Absolutely, in fact the example game included should give you a good taste of what it's capable of as a moba style game.

**Reply by cody3b** [reply]
Thanks!

**Reply by elementbound** [reply]
Let me add a few thoughts in no particular order to expand on this.

The more complex and more time-sensitive interactions you want to implement, the more difficult it will be to get it right, especially under bigger latencies. It's good to establish some baselines for your game. For example, for Forest Brawl, the guideline was to make it perform great at 50ms, and perform playable at 200ms.

Games tend to employ tricks / simplifications to keep things running. For example in some MOBAs ( e.g. LoL ), your basic attacks just "home in" on the target - this way you don't need to do lag compensation for aiming, since the projectile will always hit the target. 

Similarly, I have a strong hunch that many MOBAs employ deterministic lockstep, potentially with rollback. There is currently no support implemented in netfox, so if you'd want to go that route, you'd need to roll your own lockstep loop, similar to NetworkRollback. The good news is, deterministic lockstep is just one approach, and is not mandatory for this kind of game. It should be doable with netfox's rollback implementation as well. 

Also, due to an [engine limiation](https://foxssake.github.io/netfox/netfox/tutorials/rollback-caveats/#physics-updates), collisions are not fully up-to-date during rollback. This is fine when colliding with static stuff ( e.g. the level ), but can be less reliable when two dynamic objects collide during rollback. Due to this, I recommend running your ability logic *outside* of the rollback loop. 

Lastly, I recommend doing a tech demo of your game first. This would be a small version with placeholder graphics, that implements all the major mechanics of your game. This way, you can tackle all the networking-related issues up front and see how the netcode behaves. It also helps you get into the mindset of thinking about how your game state is replicated over the network. With some luck, you manage to implement most of your systems as reusable classes / methods /etc., and just build on top of them for the actual game. 

Best of luck, and feel free to open discussions / issues as you progress! And don't forget to [showcase](https://github.com/foxssake/netfox/discussions/207) either if you get to that point :slightly_smiling_face:
